src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -149,15 +176,355 @@
         setSystemGeneratedData(invoiceFromStorage, invoice);
         return invoiceFromStorage;
       })
-      .thenCompose(invoiceFromStorage -> {
-        if (isTransitionToPaid(invoiceFromStorage, invoice)) {
-          return payInvoice(invoice);
-        } else {
-          return updateInvoiceRecord(invoice);
+      .thenCompose(invoiceFromStorage -> handleInvoiceStatusTransition(invoice, invoiceFromStorage))
+      .thenCompose(aVoid -> updateInvoiceRecord(invoice));
+  }
+
+  private void setSystemGeneratedData(Invoice invoiceFromStorage, Invoice invoice) {
+    invoice.withFolioInvoiceNo(invoiceFromStorage.getFolioInvoiceNo());
+  }
+
+  private CompletionStage<Void> handleInvoiceStatusTransition(Invoice invoice, Invoice invoiceFromStorage) {
+    if (isTransitionToApproved(invoiceFromStorage, invoice)) {
+      return approveInvoice(invoice);
+    } else if (isTransitionToPaid(invoiceFromStorage, invoice)) {
+      return payInvoice(invoice);
+    }
+    return VertxCompletableFuture.completedFuture(null);
+  }
+
+  private boolean isTransitionToApproved(Invoice invoiceFromStorage, Invoice invoice) {
+    return invoiceFromStorage.getStatus() == Invoice.Status.REVIEWED && invoice.getStatus() == Invoice.Status.APPROVED;
+  }
+
+  /**
+   * Handles transition of given invoice to {@link Invoice.Status#APPROVED} status.
+   * Transition only triggers if the current {@link Invoice.Status} is {@link Invoice.Status#REVIEWED}
+   * and exist at least one {@link InvoiceLine} associated with this invoice
+   *
+   * @param invoice {@link Invoice}to be approved
+   * @return CompletableFuture that indicates when transition is completed
+   */
+  private CompletableFuture<Void> approveInvoice(Invoice invoice) {
+    CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+    getInvoiceLinesWithTotals(invoice)
+      .thenApply(invoiceLines -> {
+        verifyInvoiceLineNotEmpty(invoiceLines);
+        return invoiceLines;
+      })
+      .thenCombine(prepareVoucher(invoice), (invoiceLines, voucher) -> {
+        invoice.setVoucherNumber(voucher.getVoucherNumber());
```

Comment:
didn't notice that we have this two-way relationship... Making a note to see if we can avoid this.

Reply:
The voucherLines also refers twice to invoiceLines: 
- [sourceIds](https://github.com/folio-org/acq-models/blob/163af732853b1a7a4cc8e9b48f57f7f7cd7cb387/mod-invoice-storage/schemas/voucher_line.json#L28)
- [fundDistributions[].invoiceLineId](https://github.com/folio-org/acq-models/blob/163af732853b1a7a4cc8e9b48f57f7f7cd7cb387/mod-invoice-storage/schemas/fund_distribution.json#L20)

Reply:
Thanks for pointing this out